etcd/etcd-live-http-to-https-migration.md

```diff
@@ -0,0 +1,192 @@
+# etcd Live HTTP to HTTPS migration
+
+If you already have existing etcd cluster which uses insecure HTTP connections and you would like to switch from HTTP to HTTPS and in addition enable TLS certificate authentication, you have to follow instruction below.
+
+## Prepare Cluster Components
+
+### Check insecure port availability
+
+By default fleet and flannel use two etcd client ports: 2379 (IANA port assignment) and 4001 (legacy port). Last one is used for etcd v0.4 backward compatibility. In this example we will use 4001 TCP port for insecure HTTP connections and our etcd members will listen that port only on local interface for security reason. This will allow us to switch transparently to HTTPS configuration without services downtime and simplify configuration
+
+If you've configured flannel, fleetd or other local components to use custom or 2379 port only, you have to configure them to use 4001 TCP port.
+
+If your etcd configuration doesn't have configured 4001 client port, you have to enable it before we proceed with this manual. For example you use Cloud-Config to configure your CoreOS cluster. In this case you have to retrieve `ETCD_LISTEN_CLIENT_URLS` value from `/run/systemd/system/etcd2.service.d/20-cloudinit.conf` config file:
+
+```sh
+grep ETCD_LISTEN_CLIENT_URLS /run/systemd/system/etcd2.service.d/20-cloudinit.conf
+Environment="ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379"
+```
+
+Then create `/etc/systemd/system/etcd2.service.d/` directory if it doesn't exist and create [drop-in][drop-ins] file: `/etc/systemd/system/etcd2.service.d/25-insecure_localhost.conf`. Use value we've retrieved above and append `http://127.0.0.1:4001` URL at the end:
+
+```
+[Service]
+Environment="ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379,http://127.0.0.1:4001"
+```
+
+Then run `systemctl daemon-reload` and `systemctl restart etcd2` to restart etcd. Check cluster availability using these commands:
+
+```sh
+etcdctl member list
+etcdctl cluster-health
+```
+
+Repeat these steps on each CoreOS host which has etcd member.
+
+### Generate TLS keypairs
+
+Follow [generate self signed certificates][self-signed-ca] doc to generate self-signed certificates.
+
+### Copy Keypairs
+
+Let's assume we have three CoreOS hosts with three etcd cluster members: server1, server2, server3 and corresponding IP addresses: 172.16.0.101, 172.16.0.102, 172.16.0.103. We will use following keypairs filenames in our example:
+
+```
+server1.pem
+server1-key.pem
+```
+
+Copy your `ca.pem` CA certificate into `/etc/ssl/certs` directory and run `update-ca-certificates` script to update your certificates bundle.
+
+Create `/etc/etcd2` directory, copy there corresponding certificate and key and set file permissions:
+
+```
+chown -R etcd:etcd /etc/etcd2/
+chmod 600 /etc/etcd2/*-key.pem
+```
+
+Repeat this step on rest of the member nodes.
+
+### Configure etcd Proxy
```

Comment:
Add somewhere info about proxy: if you've used apps which work with etcd using insecure connection to remote host. we recommend you to to install proxy which will listen localhost and reconfigure apps to connect to localhost's insecure port. that will be much easier rather then configure each app to use certs. especially when these apps probably doesn't have certs support.